BUG,MAINT: Ensure masked elements can be tested against nan and inf.

mhvk opened the pull request:
The removal of nan and inf from arrays that are compared using test routines like assert_array_equal treated the two arrays separately, which for masked arrays meant that some elements would not be removed when they should have been. This PR corrects this.

fixes #11121

mhvk committed:
BUG,MAINT: Ensure masked elements can be tested against nan and inf.

The removal of nan and inf from arrays that are compared using
test routines like assert_array_equal treated the two arrays
separately, which for masked arrays meant that some elements would
not be removed when they should have been. This PR corrects this.

charris commented on numpy/testing/_private/utils.py:
```diff
@@ -695,9 +695,14 @@ def isnumber(x):
-    def chk_same_position(x_id, y_id, hasval='nan'):
-        """Handling nan/inf: check that x and y have the nan/inf at the same
-        locations."""
+    def func_same_pos(func, x, y, hasval='nan'):
```
Needs a full docstring. I think the original name is a bit better. Maybe change the signature to `(x, y, func, hasval='nan')`. Do we really need both `hasval` and `func`?

mhvk replied:
`func` does not always have a name (used with a `lambda` below to test for +/- inf). I'll change to `func_chk_same_pos`.

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -706,6 +711,16 @@ def chk_same_position(x_id, y_id, hasval='nan'):
+            x = x[~flag]
+        if y.shape:
+            y = y[~flag]
+        return x, y
```
I don't understand this function - can't it just be written `return x[~flag], y[~flag]`?

mhvk replied:
No, `x` or `y` can be a scalar while the other is an array. This is why the flags were treated separately before.

eric-wieser replied:
The issue here is avoiding `arr_0d[bool_nd]`, right? Would `x, y = np.broadcast_arrays(x, y)` be off the table?

mhvk replied:
Not crazy, but I'd rather stick with what I have here...

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -719,32 +734,23 @@ def chk_same_position(x_id, y_id, hasval='nan'):
+                else:
+                    x, y = remove_flagged(x, y, xy_isnan)
+            elif xy_isinf is not False:
+                x, y = remove_flagged(x, y, xy_isinf)
```
Seems like micro-optimization here - can't we just write in place of these 7 lines:
```
to_remove = xy_isinf | xy_isnan
x = x[~to_remove]
y = y[~to_remove]
```

mhvk replied:
Doesn't work for the same reason.

mhvk replied:
The problem was that `xy_isinf` could just be `False`, in which case the flags would not broadcast correctly. But of course, this is easily checked in `remove_flagged`, so I did that now.

mhvk commented:
@charris, @eric-wieser - thanks for the comments; I made some changes which I think make the use clearer.

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -706,6 +711,19 @@ def chk_same_position(x_id, y_id, hasval='nan'):
+        if flag is False:
+            return x, y
+        if flag.all():
+            return array(()), array(())
```
nit: I'd use `[]` here for a size-0 array - I'm used to seeing `()` for a 0dim array, so this is a little jarring

mhvk replied:
OK

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -706,6 +711,19 @@ def chk_same_position(x_id, y_id, hasval='nan'):
+        return x_id if x_id.shape else y_id
+
+    def remove_flagged(x, y, flag):
+        """Remove elements in x and y, noting that either may be scalar."""
```
python or numpy scalar?

mhvk replied:
numpy, I'll add to the comment.

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -695,9 +695,14 @@ def isnumber(x):
-    def chk_same_position(x_id, y_id, hasval='nan'):
-        """Handling nan/inf: check that x and y have the nan/inf at the same
-        locations."""
+    def func_chk_same_pos(x, y, func=isnan, hasval='nan'):
```
Can we spell this `check` instead of `chk`? Or even `assert_same_predicate_pos`?

mhvk replied:
OK, `assert_same_pos`

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -706,6 +711,20 @@ def chk_same_position(x_id, y_id, hasval='nan'):
                                 % (hasval), verbose=verbose, header=header,
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
+        return x_id if x_id.shape else y_id
```
I think this is backward - you should return the scalar boolean mask here, not the array one (which would at this point be equal to the larger array anyway). That allows you to remove all of your special cases in `remove_flagged`

mhvk replied:
Yes, indeed, you are completely right - that gets rid of `remove_flagged` altogether. Now implemented.

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -719,32 +738,19 @@ def chk_same_position(x_id, y_id, hasval='nan'):
-                x = x[~x_isinf]
-                y = y[~y_isinf]
+                xy_isinf = func_assert_same_pos(x, y,
+                                                func=lambda xy: xy == +inf,
```
`np.isposinf`?

mhvk replied:
Learn something new every day...

mhvk replied:
But doesn't work for complex numbers...

eric-wieser replied:
Does what you have here work for complex numbers?

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -706,6 +711,22 @@ def chk_same_position(x_id, y_id, hasval='nan'):
+        # flag as it everywhere, so we should return the scalar flag.
+        return x_id if not x_id.shape else y_id
+
+    def remove_flagged(x, y, flag):
```
Don't forget to remove this function now

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -718,49 +739,32 @@ def chk_same_position(x_id, y_id, hasval='nan'):
-                if any(x_isnat) or any(y_isnat):
-                    x = x[~x_isnat]
-                    y = y[~y_isnat]
+        if getattr(flagged, 'shape', ()):
```
Isn't this attribute guaranteed to exist?

mhvk replied:
No, it can be `False` -- initialized as such. I tried changing, but cannot get rid of the `if flagged` stanza, nor of the `isinstance` - I suggest we leave this to another PR...

eric-wieser replied:
Just initialize it to `flagged = np.False_`?

mhvk replied:
I thought about that, but I do feel that I'm then decreasing the clarity of the rest of the code - at least in my emacs, `False` is coloured very nicely, and `false_` is not.

eric-wieser replied:
I'd argue that keeping a variable type consistent is more important than having the color look like in your editor.

> I tried changing, but cannot get rid of the if flagged stanza

I'm gonna do a checkout to see what I'm missing here

eric-wieser replied:
How about `bool_(False)`, which meets both requirements?

eric-wieser replied:
Ok, I see why the 0d case still has special handling - I'd forgotten that `arr[bool_0d].shape == (?,) + arr.shape`, and is not 1D.

eric-wieser replied:
Or an alternate spelling - `np.ndim(flagged) == 0`

mhvk replied:
Yes, I like the `ndim` - that is much clearer why we do it. Pushed up.

eric-wieser commented:
Does the inf-checking behave on complex numbers?

mhvk commented:
On the `inf` checking on complex numbers: I really don't want to go there right now...  Am having too much fun with the `gufunc` signature stuff..

eric-wieser commented:
Complex number checking is out of scope here - that's fine. Maybe open an issue about it if you can produce a test-case

mhvk commented:
Cannot immediately produce one. Partially because of an oddity:
```
1.+1j*np.inf
# (nan+infj)
```
But even accounting for that, no, I cannot seem to break it!

eric-wieser commented:
Interesting oddity there. I suppose you have to use `complex(1, np.inf)`

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -706,6 +711,9 @@ def chk_same_position(x_id, y_id, hasval='nan'):
             raise AssertionError(msg)
+        # If there is a scalar, then here we know the array has the same
+        # flag as it everywhere, so we should return the scalar flag.
+        return x_id if not x_id.shape else y_id
```
nit: I think this is clearer as `x_id.ndim == 0` rather than `not x_id.shape`

mhvk replied:
Yes, done.

eric-wieser commented:
Just the nits about `ndim` and `shape` remain - you're right that it's harder to remove the special cases than I'd though

eric-wieser commented:
Took the liberty of adding and explicit `!= 0` comparison, since that's normally how we spell ndim comparisons

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -695,9 +695,14 @@ def isnumber(x):
+        x_id = func(x)
+        y_id = func(y)
+        if not any(x_id) and not any(y_id):
+            return False
```
Are these two lines needed? Aren't they covered by the below?

mhvk replied:
You'd think not, but numerous recursion failures if I leave them out...

mhvk replied:
So, what happens is that it does recurse, but the `if` statement prevents it from doing it more than once.

mhvk replied:
which is simple to solve - just use `any` - will push fix up shortly.

mhvk replied:
Doesn't work either, but passing in `equal_[nan|inf]=False` does.

mhvk replied:
No, it doesn't...

eric-wieser replied:
Maybe add a comment saying that this is needed to prevent recursion or something?

Good to go either way

mhvk committed:
MAINT: clean up assert_array_compare a bit further.

This brought to light two bugs in tests, which are fixed here, viz.,
that a sample ndarray subclass that tested propagation of an added
parameter was incomplete, in that in propagating the parameter in
__array_wrap__ it assumed it was there on self, but that assumption
could be broken when a view of self was taken (as is done by
x[~flagged] in the test routine), since there was no
__array_finalize__ defined.

The other subclass bug counted, incorrectly, on only needing to provide
one type of comparison, the __lt__ being explicitly tested.  But flags
are compared with __eq__ and those flags will have the same subclass.

mhvk commented:
What a pain! But finally got it working with a more sensible implementation. Turns out I was partially misled by test that failed because they used subclasses that were not written correctly (one by me...).

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -695,17 +695,23 @@ def isnumber(x):
+    def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
+        """Handling nan/inf: combine results of running func on x and y,
+        checking that they are True at the same locations."""
+        # Both the cast to bool_ and the != True comparison are done
```
Can you move the cast to bool to the scalar case below?

mhvk replied:
I needed to cast both `x_id` and `y_id`, but that can indeed be done at the end. That does look a bit nicer, so done.

eric-wieser commented on numpy/lib/tests/test_ufunclike.py:
```diff
@@ -55,6 +55,10 @@ def __array_wrap__(self, obj, context=None):
                 return obj
 
+            def __array_finalize__(self, obj):
+                self.metadata = getattr(obj, 'metadata', None)
```
Can you describe in the commit message why these changes were needed?

mhvk replied:
Done. This was indeed a non-trivial one - I had to look again to remind me why it was needed.  I wondered about the use of stating it in the commit message until I recalled that with `git blame` one can find why this was added.

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -695,17 +695,24 @@ def isnumber(x):
+        # If there is a scalar, then here we know the array has the same
+        # flag as it everywhere, so we should return the scalar flag.
+        return (bool_(x_id) if x_id.ndim == 0 else
+                (bool_(y_id) if y_id.ndim == 0 else y_id))
```
Probably clearer as just a normal if / else chain at this point.

mhvk replied:
Agreed; done.

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -718,49 +725,32 @@ def chk_same_position(x_id, y_id, hasval='nan'):
+            if x.size == 0:
+                return
+        elif flagged:
+            # no sense doing comparison if everything is flagged.
```
Not sure I agree with this - what if `x` and `y` are quantities of different dimensions? The test should still compare the empty arrays so that the correct error is raised, right?

mhvk replied:
It is the same case as the `if x.size == 0` stanza above, but for the case where `x` and `y` are both scalar. In principle, I agree the comparison should work for empty elements, but in practice, this is not the case (e.g., `matrix` changing its shape being one problem...)

Since this was here before, I propose to leave it as is.

eric-wieser replied:
Sure, i agree this is out of scope for the patch

mhvk replied:
Overall, it sadly remains a chunk of rather fragile code...

eric-wieser commented:
Sorry for dropping the ball on this one. Would be great to get this into 1.15.

charris commented:
Thanks Marten.

charris commented on numpy/testing/_private/utils.py:
```diff
@@ -695,17 +695,28 @@ def isnumber(x):
+        # compared usefully, and for which .all() yields masked.
+        x_id = func(x)
+        y_id = func(y)
+        if (x_id == y_id).all() != True:
```
Hmm,
```
In [10]: ma.masked != True
Out[10]: masked

In [11]: ma.masked == True
Out[11]: masked

In [12]: bool(ma.masked)
Out[12]: False
```

mhvk replied:
The purpose is to ensure that the stanza is *not* executed if `x_id` or `y_id` is `masked`, yet is executed if they are different.
